modules/dnn/src/webgpu/src/buffer.cpp

```diff
@@ -0,0 +1,55 @@
+#include "../include/buffer.hpp"
+#include "../dawn/dawnUtils.hpp"
+#include "common.hpp"
+#include <string.h>
+namespace cv { namespace dnn { namespace webgpu {
+// #ifdef HAVE_WEBGPU
+
+Buffer::Buffer(std::shared_ptr<wgpu::Device> device)
+{
+    device_ =  device;
+    usage_ = wgpu::BufferUsage::Storage;
+    bufferMapped_->buffer = nullptr;
+    bufferMapped_->dataLength = 0;
+    bufferMapped_->data = nullptr;
+}
+
+Buffer::Buffer( std::shared_ptr<wgpu::Device> device, 
+                const void* data, size_t size, 
+                wgpu::BufferUsage usage) 
+{
+    device_ = device;
+    usage_ = usage;
+    size_ = size;
+    wgpu::BufferDescriptor descriptor = {};
+    descriptor.size = size;
+    descriptor.usage = usage;
+    bufferMapped_ = std::make_shared<wgpu::CreateBufferMappedResult>
+                    (device_->CreateBufferMapped(&descriptor));
+    if(data) { memcpy(bufferMapped_->data, data, size); }
+}
+
+const void* Buffer::MapReadAsyncAndWait() 
```

Comment:
As my understanding, no need to create a new buffer and copy the content here. 
Just use wgpu::Buffer::MapReadAsync() and condition variable to make the map process from async to sync.